### PR TITLE
Loki microservices Tanka: fix error during manifest generation

### LINKED ISSUE #9361: Loki microservices Tanka: error during manifest generation
**Describe the bug**
There is an error during manifest generation with the following options:
```
memberlist_ring_enabled: true,
create_service_monitor: true,
```

The error is:
`Error: evaluating jsonnet: RUNTIME ERROR: Field does not exist: name
        /home/ivan.rizzante/kubernetes/gitops/loki/vendor/loki/servicemonitor.libsonnet:42:13-42        thunk from <object <anonymous>>
        Array element 3
        Field "values"
        Array element 0
        Field "matchExpressions"
        Field "selector"
        Field "spec"
        Field "serviceMonitor"
        During manifestation`

**To Reproduce**
1. Started Loki 2.8.1
2. Set configuration parameters as above
3. Render jsonnet manifests with `tk show`

**Expected behavior**
I should be able to generate the manifests

**Environment:**
 - Infrastructure: kubernetes
 - Deployment tool: jsonnet

**Screenshots, Promtail config, or terminal output**
If applicable, add any output to help explain your problem.

### PR DESCRIPTION
**Pull request by irizzant**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:
Fixes #9361 

This PR fixes an error on the ServiceMonitor generated when
```
memberlist_ring_enabled: true,
create_service_monitor: true,
```
It also adds some checks on `memberlist.libsonnet` to avoid generating invalid kubernetes manifests.

**Special notes for your reviewer**:

**Checklist**
- [X] Reviewed the [`CONTRIBUTING.md`](https://github.com/grafana/loki/blob/main/CONTRIBUTING.md) guide (**required**)
- [X] Documentation added
- [X] Tests updated
- [X] `CHANGELOG.md` updated
- [X] Changes that require user attention or interaction to upgrade are documented in `docs/sources/upgrading/_index.md`

### CODE REVIEW

**Comment by irizzant** [comment]
@DylanGuedes again? ;)

File: production/ksonnet/loki/servicemonitor.libsonnet [hunk 1]
```diff
@@ -41,7 +41,7 @@
           values: [
             this[name].spec.selector.name
             for name in std.objectFields(this)
-            if std.isObject(this[name]) && std.objectHas(this[name], 'kind') && this[name].kind == 'Service'
+            if std.isObject(this[name]) && std.objectHas(this[name], 'kind') && this[name].kind == 'Service' && std.objectHasAll(this[name].spec.selector, 'name')
```
Comment (DylanGuedes) [inline production/ksonnet/loki/servicemonitor.libsonnet:44]: could you confirm if it shouldn't be `name` instead of `'name'`?

**Review by DylanGuedes — CHANGES_REQUESTED** [review]
Thanks for your contribution! Added a question

**Comment by irizzant** [comment]
@DylanGuedes do you need anything else for this to be merged?

**Comment by DylanGuedes** [comment]
> @DylanGuedes do you need anything else for this to be merged?

i asked you [this question](https://github.com/grafana/loki/pull/9362#discussion_r1182331878).

**Comment by irizzant** [comment]
@DylanGuedes sorry I thought you saw this reply:
image

**Comment by DylanGuedes** [comment]
> @DylanGuedes sorry I thought you saw this reply: image

it won't show up for me because it is pending :rofl: anyway, approved

**Pull request merged**